azure-iot-device/tests/common/pipeline/test_pipeline_stages_base.py

```diff
@@ -291,12 +289,150 @@ def test_already_connected(self, mocker, stage, op):
         assert stage.send_op_down.call_count == 1
         assert stage.send_op_down.call_args == mocker.call(op)
 
+    @pytest.mark.it("Allows the op to complete successfully")
+    def test_success(self, mocker, stage, op):
+        callback = op.callback_stack[0]
+        stage.pipeline_root.connected = True
+
+        stage.run_op(op)
+
+        # make sure our call went down
+        assert stage.send_op_down.call_count == 1
+        assert stage.send_op_down.call_args[0][0] is op
+
+        # complete the op
+        op.complete()
+
+        # assert that the op completed successfully
+        assert op.complete
+        assert op.error is None
+        assert callback.call_count == 1
+        assert callback.call_args == mocker.call(op=op, error=None)
+
     @pytest.mark.it(
-        "Sends a new ConnectOperation down the pipeline if the pipeline is not yet in a 'connected' state"
+        "Does not complete the operation if the op fails and the transport is no longer connected"
     )
+    def test_failure_and_disconnect_no_complete(self, mocker, stage, op, arbitrary_exception):
+        stage.pipeline_root.connected = True
+
+        stage.run_op(op)
+
+        # make sure our call went down
+        assert stage.send_op_down.call_count == 1
+        assert stage.send_op_down.call_args[0][0] is op
+
+        # disconnect the transport and complete the op with failure
+        stage.pipeline_root.connected = False
+        op.complete(arbitrary_exception)
+
+        # assert that the op didn't complete
+        assert not op.completed
+
+    @pytest.mark.it(
+        "Sends down a new ConnectOperation if the op fails and the transport is no longer connected"
+    )
+    def test_failure_and_disconnect_new_connect_operation(
+        self, mocker, stage, op, arbitrary_exception
+    ):
+        stage.pipeline_root.connected = True
+        stage.run_op(op)
+
+        # make sure our call went down
+        assert stage.send_op_down.call_count == 1
+        assert stage.send_op_down.call_args[0][0] is op
+
+        # disconnect the transport and complete the op with failure
+        stage.pipeline_root.connected = False
+        op.complete(arbitrary_exception)
+
+        # assert that we got a new connect operation
+        assert stage.send_op_down.call_count == 2
+        connect_op = stage.send_op_down.call_args[0][0]
+        assert isinstance(connect_op, pipeline_ops_base.ConnectOperation)
+
+    @pytest.mark.it(
+        "Runs the operation a second time if the stage needs to reconnect after the op fails"
+    )
```

Comment:
Could be good to have a test to show that this works more than just twice, i.e. n times, but not a huge deal